src/plugins/crypto/gcrypt_operations.c

```diff
@@ -0,0 +1,221 @@
+		long contentLen = ELEKTRA_CRYPTO_GCRY_BLOCKSIZE;
+
+		if((i + 1) * ELEKTRA_CRYPTO_GCRY_BLOCKSIZE > valueLen)
+		{
+			contentLen = valueLen - (i * ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+		}
+		memcpy(contentBuffer, (value + i), contentLen);
+		if(contentLen < ELEKTRA_CRYPTO_GCRY_BLOCKSIZE)
+		{
+			elektraCryptoAddPkcs7Padding(contentBuffer, contentLen, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+		}
+
+		gcry_err = gcry_cipher_encrypt(gcry_handle, cipherBuffer, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE, contentBuffer, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+		if(gcry_err != 0)
+		{
+			// TODO forward detailed error description with gcry_strerror() and gcry_strsource()
+			free(output);
+			return ELEKTRA_CRYPTO_GCRY_NOK;
+		}
+		memcpy((output + i), cipherBuffer, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+	}
+
+	// write back the cipher text to the key
+	keySetBinary(k, output, outputLen);
+	free(output);
+
+	return ELEKTRA_CRYPTO_GCRY_OK;
+}
+
+int elektraCryptoGcryDecrypt(Key *k)
+{
+	const unsigned char *value = (unsigned char*)keyValue(k);
+	const size_t valueLen = keyGetValueSize(k);
+
+	unsigned char *output;
+	unsigned char cipherBuffer[ELEKTRA_CRYPTO_GCRY_BLOCKSIZE];
+	unsigned char contentBuffer[ELEKTRA_CRYPTO_GCRY_BLOCKSIZE];
+	unsigned long i;
+	unsigned long written = 0;
+	unsigned long lastBlockLen;
+
+	// plausibility check
+	if(valueLen % ELEKTRA_CRYPTO_GCRY_BLOCKSIZE != 0)
+	{
+		return ELEKTRA_CRYPTO_GCRY_NOK;
+	}
+
+	// prepare buffer for plain text output
+	output = (unsigned char*)malloc(valueLen);
+	if(output == NULL)
+	{
+		return ELEKTRA_CRYPTO_GCRY_NOK;
+	}
+
+	// decrypt content block by block (i = start of the current block)
+	for(i = 0; i < valueLen; i += ELEKTRA_CRYPTO_GCRY_BLOCKSIZE)
+	{
+		// load cipher text partition into the cipher buffer
+		memcpy(cipherBuffer, (value + i), ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+
+		gcry_err = gcry_cipher_decrypt(gcry_handle, contentBuffer, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE, cipherBuffer, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+		if(gcry_err != 0)
+		{
+			// TODO forward detailed error description with gcry_strerror() and gcry_strsource()
+			free(output);
+			return ELEKTRA_CRYPTO_GCRY_NOK;
+		}
+		memcpy((output + i), contentBuffer, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+		written += ELEKTRA_CRYPTO_GCRY_BLOCKSIZE;
+	}
+
+	// consider that the last block may contain a PKCS#7 padding
+	lastBlockLen = elektraCryptoGetPkcs7PaddedContentLen((output + written - ELEKTRA_CRYPTO_GCRY_BLOCKSIZE) , ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+	if(lastBlockLen < ELEKTRA_CRYPTO_GCRY_BLOCKSIZE)
+	{
+		written = written - (ELEKTRA_CRYPTO_GCRY_BLOCKSIZE - lastBlockLen);
+	}
+
+	// write back the cipher text to the key
+	// TODO consider that the keySetString() function should be applied if the original value was of type string
+	keySetBinary(k, output, written);
+	free(output);
+
+	return ELEKTRA_CRYPTO_GCRY_OK;
+}
+
+void elektraCryptoAddPkcs7Padding(unsigned char *buffer, const unsigned int contentLen, const unsigned int bufferLen)
```

Comment:
Some explanation to method please. And is such low-level stuff really needed in crypto-API?
(internal stuff should be static)